### ISSUE TITLE
Specialize error messages for attribute access on TypedDicts

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/pyright#7579]
## Background

A developer wrote some sample code like the following:

[Pyright Playground Link](https://pyright-play.net/?strict=true&enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjAFC1UA2AhgM6tQAKxIrYKACiIIylGgEoAXLSiyoKZhGKSorGCHqliwKMACuKKgIQruvflJlyEAOgVKgA)

```py
from typing import TypedDict

class Person(TypedDict):
    name: str

def func(p: Person):
    p.name
```

I ran this in strict checking mode and the code resulted in a pretty opaque set of errors:

```
Cannot access member "name" for type "Person"
  Member "name" is unknown  (reportAttributeAccessIssue)

Type of "name" is unknown  (reportUnknownMemberType)
```

I was momentarily confused, but then realized they had made a really common mistake - writing `p.name` instead of `p["name"]`. I've done this myself a bunch.

## Suggestion

It would be great if Pyright could detect a failed lookup with a corresponding key, and provide a specialized error message like

```
Cannot access member "name" in type "Person", but a key with the same name exists on this TypedDict.
```

or

```
Type "Person" is a TypedDict and does not have a member named "name". It is likely that you meant to write an indexing expression.
```

or something similar.

I would also suggest that Python developers *not* using type-checking might benefit from these as a suggestion diagnostic.

In either case, you can provide a quick fix even if it is relatively simple.

## Precedence

In TypeScript, we provide lots of error messages like this.

### Spelling Mistakes

```ts
function f(x: string) {
  x.touppercase
}
```

```
Property 'touppercase' does not exist on type 'string'. Did you mean 'toUpperCase'?(2551)

(Related info: 'toUpperCase' is declared here.)
```

[Playground Link](https://www.typescriptlang.org/play?#code/MYGwhgzhAEDC0G8BQ1oGIIHsC2BTACgE4CWAbmAC67QC80AjAEwDMA3EitAGaaYAUASkSdUFABbEIAOix4iZSrk4BfJKqA)

This one also appears as a suggestion diagnostic for JavaScript users in VS Code. That way we can provide quick fixes.

```js
/** @type {(x: string) => void} */
function foo(x) {
    x.touppercase
}
```

![Error message accessing incorrectly-spelled toUpperCase off of a string in JS](https://github.com/microsoft/pyright/assets/972891/88a3221c-542e-4d7b-9755-3762c23a11cd)

![Quick fix to correct spelling of toUpperCase](https://github.com/microsoft/pyright/assets/972891/add97d54-bbe1-40b8-b31f-8ef00c3dd9a3)

### Indexed Access Types Written as Qualified Names

[Playground Link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgGIHt3IN4ChnIDO6AthAApToAO0YAngFxFhSgDmA3LgL664AbCGCKlhACw7MM6AHTEylGnXpA)

```ts
interface Foo {
  someProperty: string;
}

let something: Foo.someProperty
```

```
Cannot access 'Foo.someProperty' because 'Foo' is a type, but not a namespace. Did you mean to retrieve the type of the property 'someProperty' in 'Foo' with 'Foo["someProperty"]'?
```

### Tuple Labels Written as Property Accesses

We haven't implemented this one yet, though it is pretty close to what's being recommended here.

https://github.com/microsoft/TypeScript/issues/56825

**Comment by erictraut** [comment]
Thanks for the suggestion. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.357, which I just published.
